cmd/remediation-aggregator/main.go

```diff
@@ -189,36 +191,97 @@ func annotateParsedConfigMap(clientset *kubernetes.Clientset, cm *v1.ConfigMap)
+
+	err := backoff.Retry(func() error {
+		err := crClient.client.Create(context.TODO(), res)
+		if err != nil && !errors.IsAlreadyExists(err) {
+			fmt.Printf("Retrying with a backoff because of an error: %v", err)
 			return err
 		}
+		return nil
+	}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetries))
+	if err != nil {
+		fmt.Printf("Failed to create a '%s' object: %v\n", kind.GroupVersionKind().Kind, err)
+		return err
+	}
+	return nil
+}
+
+func getRemediationLabels(scan *compv1alpha1.ComplianceScan) map[string]string {
+	labels := make(map[string]string)
+	labels[compv1alpha1.ScanLabel] = scan.Name
+	labels[compv1alpha1.SuiteLabel] = scan.Labels["compliancesuite"]
+	labels[mcfgv1.MachineConfigRoleLabelKey] = utils.GetFirstNodeRole(scan.Spec.NodeSelector)
+	if labels[mcfgv1.MachineConfigRoleLabelKey] == "" {
+		fmt.Printf("scan %s has no role assignment", scan.Name)
+		return nil
+	}
 
-		if rem.Labels == nil {
-			rem.Labels = make(map[string]string)
+	return labels
+}
+
+func getCheckResultLabels(scan *compv1alpha1.ComplianceScan) map[string]string {
+	labels := make(map[string]string)
+	labels[compv1alpha1.ScanLabel] = scan.Name
+	labels[compv1alpha1.SuiteLabel] = scan.Labels["compliancesuite"]
+
+	return labels
+}
+
+func createResults(crClient *complianceCrClient, scan *compv1alpha1.ComplianceScan, parsedResults []*utils.ParseResult) error {
+	fmt.Printf("Will create %d result objects\n", len(parsedResults))
+	if len(parsedResults) == 0 {
+		fmt.Println("Nothing to create")
+		return nil
+	}
+
+	for _, pr := range parsedResults {
+		if pr == nil || pr.CheckResult == nil {
+			fmt.Println("nil result or result.check, this shouldn't happen")
+			continue
 		}
-		rem.Labels[compv1alpha1.ScanLabel] = scan.Name
-		rem.Labels[compv1alpha1.SuiteLabel] = scan.Labels["compliancesuite"]
-		rem.Labels[mcfgv1.MachineConfigRoleLabelKey] = utils.GetFirstNodeRole(scan.Spec.NodeSelector)
-		if rem.Labels[mcfgv1.MachineConfigRoleLabelKey] == "" {
-			return fmt.Errorf("scan %s has no role assignment", scan.Name)
+
+		checkResultLabels := getCheckResultLabels(scan)
+		if checkResultLabels == nil {
+			return fmt.Errorf("cannot create checkResult labels")
 		}
 
-		err := backoff.Retry(func() error {
-			err := crClient.client.Create(context.TODO(), rem)
-			if err != nil && !errors.IsAlreadyExists(err) {
-				return err
-			}
-			return nil
-		}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetries))
-		if err != nil {
+		// check is owned by the scan
+		if err := createOneResult(crClient, scan, checkResultLabels, pr.CheckResult); err != nil {
+			return fmt.Errorf("cannot create checkResult %s: %v", pr.CheckResult.Name, err)
+		}
 
-			fmt.Printf("Failed to create remediation: %v\n", err)
-			return err
+		if pr.Remediation == nil {
+			continue
 		}
+
+		remLabels := getRemediationLabels(scan)
+		if remLabels == nil {
+			return fmt.Errorf("cannot create remediation labels")
```

Comment:
This is gonna be really hard to figure out when it shows up in the logs. Let's instead return an error from the `getRemediatoinLabels` function, which will say exactly what happened.

Reply:
fixed locally